File: autocontext/src/autocontext/loop/annealing.py
from __future__ import annotations

import math
import random
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True, slots=True)
class AnnealingSchedule:
    enabled: bool = False
    start_temperature: float = 0.05
    end_temperature: float = 0.001
    generations: int = 20


@dataclass(frozen=True, slots=True)
class AnnealingOutcome:
    accepted: bool
    temperature: float
    acceptance_probability: float
    random_value: float
    delta: float

    def model_dump(self) -> dict[str, Any]:
        return {
            "accepted": self.accepted,
            "temperature": self.temperature,
            "acceptance_probability": self.acceptance_probability,
            "random_value": self.random_value,
            "delta": self.delta,
        }


def annealing_temperature(schedule: AnnealingSchedule, generation: int) -> float:
    span = max(1, schedule.generations - 1)
    progress = min(1.0, max(0.0, (generation - 1) / span))
    return max(0.0, schedule.start_temperature + (schedule.end_temperature - schedule.start_temperature) * progress)


def annealing_random_value(seed_base: int, generation: int, attempt: int) -> float:
    return random.Random(seed_base + generation * 10_000 + attempt).random()


def evaluate_annealing(delta: float, schedule: AnnealingSchedule, generation: int, random_value: float) -> AnnealingOutcome:
    temperature = annealing_temperature(schedule, generation)
    probability = math.exp(delta / temperature) if schedule.enabled and delta < 0 and temperature > 0 else 0.0
    probability = min(1.0, max(0.0, probability))
    return AnnealingOutcome(
        accepted=schedule.enabled and delta < 0 and random_value < probability,
        temperature=temperature,
        acceptance_probability=probability,
        random_value=random_value,
        delta=delta,
    )

File: ts/src/loop/annealing.ts
import type { GateDecision } from "./backpressure.js";

export interface AnnealingOptions {
  enabled?: boolean;
  generation: number;
  randomValue: number;
  startTemperature?: number;
  endTemperature?: number;
  generations?: number;
}

export interface AnnealingOutcome {
  accepted: boolean;
  temperature: number;
  acceptanceProbability: number;
  randomValue: number;
  delta: number;
}

export function annealingTemperature(opts: AnnealingOptions): number {
  const start = opts.startTemperature ?? 0.05;
  const end = opts.endTemperature ?? 0.001;
  const generations = opts.generations ?? 20;
  const span = Math.max(1, generations - 1);
  const progress = Math.min(1, Math.max(0, (opts.generation - 1) / span));
  return Math.max(0, start + (end - start) * progress);
}

export function deterministicAnnealingRandomValue(
  seedBase: number,
  generation: number,
  attempt: number,
): number {
  let hash = 0x811c9dc5;
  const key = `${seedBase}:${generation}:${attempt}`;
  for (let index = 0; index < key.length; index += 1) {
    hash = Math.imul(hash ^ key.charCodeAt(index), 0x01000193) >>> 0;
  }
  return hash / 0x100000000;
}

export function evaluateAnnealing(delta: number, opts: AnnealingOptions): AnnealingOutcome {
  const temperature = annealingTemperature(opts);
  const probability = opts.enabled && delta < 0 && temperature > 0
    ? Math.min(1, Math.max(0, Math.exp(delta / temperature)))
    : 0;
  return {
    accepted: Boolean(opts.enabled && delta < 0 && opts.randomValue < probability),
    temperature,
    acceptanceProbability: probability,
    randomValue: opts.randomValue,
    delta,
  };
}

export function applyAnnealingToGateDecision(
  decision: GateDecision,
  opts: AnnealingOptions,
): GateDecision {
  if (!opts.enabled) return decision;

  const outcome = evaluateAnnealing(decision.delta, opts);
  const metadata = {
    ...decision.metadata,
    annealing: {
      accepted: outcome.accepted,
      temperature: outcome.temperature,
      acceptance_probability: outcome.acceptanceProbability,
      random_value: outcome.randomValue,
      delta: outcome.delta,
    },
  };

  if (decision.decision === "advance" || !outcome.accepted) return { ...decision, metadata };
  return {
    ...decision,
    decision: "advance",
    reason: `${decision.reason}; annealing accepted regression`,
    metadata,
  };
}
